src/analyses/callstringAnalysis.ml

```diff
@@ -0,0 +1,87 @@
+open Analyses
+open GoblintCil
+open GobConfig
+
+(* Specifies the type of the callstack elements for the callstring analysis*)
+module type Callstack_Type =
+sig
+  include CilType.S
+  val stackTypeName: string
+  val pushElem: fundec -> ('d,'g,'c,'v) ctx -> t option (* returns an elements that should be pushed to the Callstack *)
+end
+
+(** Analysis with the k-callsting approach, which uses the last k callstack elements as context.
+    With the CT argument it is possible to specify the type of the callstack elements*)
+module Spec (CT:Callstack_Type) : MCPSpec = 
+struct
+  include Analyses.IdentitySpec
+
+  (* simulates a call stack of depth k*)
+  module CallStack = struct
+    include Printable.PQueue (CT)
+    let depth = get_int "ana.context.callStack_height" (* must be >= 0 *)
+
+    let push stack elem = (* pushes elem to the stack, guarantees stack depth of k*)
+      match elem with
+      | None -> stack
+      | Some e -> 
+        let new_stack = BatDeque.cons e stack in (* pushes new element to stack*)         
+        (* removes element from stack, if stack was filled with k elements*)
+        match (BatDeque.size new_stack - depth) with
+        | x when x <= 0 -> new_stack
+        | 1 -> fst @@ Option.get (BatDeque.rear new_stack)
+        | _ -> failwith "Callstack Error: It shouldn't happen that more than one element must be deleted to maintain the correct height!"
+  end
+
+  module D = Lattice.Flat (CallStack)
+  module C = CallStack
+  module V = EmptyV
+  module G = Lattice.Unit
+
+  let name () = "callstring_"^ CT.stackTypeName
+  let startstate v = `Lifted (BatDeque.empty)
+  let exitstate v =  `Lifted (BatDeque.empty) (*TODO: should I use startstate here? Does this make a difference*)
+
+  let context fd x = match x with 
+    | `Lifted x -> x
+    | _ -> failwith "Callstring: Context error! The context cannot be derived from Top or Bottom!" (* TODO: is there a possibility???*)
+
+  let callee_state ctx f = 
+    let elem = CT.pushElem f ctx in
+    let new_stack = CallStack.push (context f ctx.local) elem in (*TODO: is it ok to use context here??? *)
+    `Lifted new_stack
+
+  let enter ctx r f args = [ctx.local, callee_state ctx f]
+
+  let combine_env ctx lval fexp f args fc au f_ask = ctx.local
+
+  let threadenter ctx ~multiple lval v args = [callee_state ctx (Cilfacade.find_varinfo_fundec v)]
```

Comment:
is this right here? Should a new thread not start with an empty callstack?

Reply:
Discussed this with @jerhard yesterday, seems like there's an argument to be made for either choice.